include/revng/Pipeline/PathTargetBimap.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+#include <tuple>
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/YAMLTraits.h"
+
+#include "revng/Pipeline/Target.h"
+#include "revng/TupleTree/TupleTreePath.h"
+
+namespace pipeline {
+class Context;
+
+class SerializableTargetInStep {
+public:
+  std::string Content;
+  std::string StepName;
+  std::string ContainerName;
+};
+
+class TargetInStep {
+private:
+  Target Content;
+  std::string StepName;
+  std::string ContainerName;
+
+public:
+  TargetInStep(const Target &Target,
+               const std::string &StepName,
+               const std::string &ContainerName) :
+    Content(Target), StepName(StepName), ContainerName(ContainerName) {}
+
+  static llvm::Expected<llvm::SmallVector<TargetInStep, 2>>
+  fromParsable(const Context &Ctx, const SerializableTargetInStep &Target);
+
+  const Target &getTarget() const { return Content; }
+
+  llvm::StringRef getStepName() const { return llvm::StringRef(StepName); }
+
+  llvm::StringRef getContainerName() const {
+    return llvm::StringRef(ContainerName);
+  }
+
+  Target &getTarget() { return Content; }
+
+  std::string &getStepName() { return StepName; }
+
+  std::string &getContainerName() { return ContainerName; }
+
+  bool operator==(const TargetInStep &) const = default;
+
+  bool operator!=(const TargetInStep &) const = default;
+
+  bool operator<(const TargetInStep &Other) const {
+    const auto &LHS = std::tie(Content, StepName, ContainerName);
+    return LHS < std::tie(Other.Content, Other.StepName, Other.ContainerName);
+  }
+
+  SerializableTargetInStep toParsable() const;
+};
+
+class PathSerializationPrimitives {
+public:
+  PathSerializationPrimitives() = default;
+
+  template<TupleTreeCompatibleAndVerifiable T>
+  static PathSerializationPrimitives make() {
+    PathSerializationPrimitives Primitives;
+    Primitives.Deserialize = &stringAsPath<T>;
+    Primitives.Serialize = &pathAsString<T>;
+    return Primitives;
+  }
+
+  std::optional<TupleTreePath> (*Deserialize)(llvm::StringRef);
+  std::optional<std::string> (*Serialize)(const TupleTreePath &);
+};
+
+class SerializableReadPathCache {
```

Comment:
A cache is something one introduces for performance reasons containing information that could be otherwise recomputed.

This is not information we can recompute, I'd call it `ReadPathsSet` or if, in principle, it doesn't need to be used only for *read* data I'd just call it `TupleTreePathsSet` or `PathsSet`.

Reply:
i went for ContainerInvalidationMetadata, as agreed